File: Study Plan/LeetCode 75/2390 Removing Stars From a String/RemovingStarsFromAString.cs
/*
    Problem
    --------------------------------------------------------------------------------
    2390. Removing Stars From a String
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 84 ms (Beats 92.86% of users with C#)
    - Memory 46.41 MB (Beats 26.25% of users with C#)
*/

public class Solution
{
    public string RemoveStars(string s)
    {
        // Hint 2
        // Use a stack to store the characters.
        var characters = new Stack<char>();

        foreach (var character in s)
        {
            if (character == '*')
            {
                // Pop one character off the stack at each star.
                characters.Pop();
            }
            else
            {
                // Otherwise, we push the character onto the stack.
                characters.Push(character);
            }
        }
        // Due to the last-in, first-out (LIFO) nature of the Stack,
        // we have to reverse it in order to restore the original sequence. 
        return new string(characters.Reverse().ToArray());
    }
}

File: Study Plan/LeetCode 75/2390 Removing Stars From a String/RemovingStarsFromAString.java
/*
    Problem
    --------------------------------------------------------------------------------
    2390. Removing Stars From a String
*/

/*  
    Solution
    --------------------------------------------------------------------------------
    - Runtime 40 ms (Beats 71.64% of users with Java)
    - Memory 47.28 MB (Beats 25.43% of users with Java)
*/

class Solution {
    public String removeStars(String s) {
        // Hint 2
        // Use a stack to store the characters.
        // https://www.baeldung.com/java-deque-vs-stack
        Deque<Character> characters = new ArrayDeque<>(s.length());
        for (Character character : s.toCharArray()) {
            if (!characters.isEmpty() && character == '*') {
                // Pop one character off the stack at each star.
                characters.pop();
            } else {
                // Otherwise, we push the character onto the stack.
                characters.push(character);
            }
        }
        StringBuilder original = new StringBuilder(characters.size());
        while (!characters.isEmpty()) {
            original.append(characters.pop());
        }
        // Due to the last-in, first-out (LIFO) nature of the Deque,
        // we have to reverse it in order to restore the original sequence. 
        return original.reverse().toString();
    }
}